src/osd/ReplicatedBackend.cc

```diff
@@ -1656,6 +1675,8 @@ int ReplicatedBackend::send_pull_legacy(int prio, pg_shard_t peer,
 
 void ReplicatedBackend::submit_push_data(
   ObjectRecoveryInfo &recovery_info,
+  ObjectRecoveryProgress &before_progress,
+  ObjectRecoveryProgress &after_progress,
```

Comment:
The whole point of passing in all of those other flags was *not* to pass in the before and after progress values.  Also, these don't appear to be modified in the below method, so you should have made them const.

I don't really want before/after_progress passed in here.  I'd rather specific arguments were added to submit_push_data computed from before_progress and after_progress before they are passed in.  It seems to me that you only need a few things:
1) interval_set regions_to_preserve (only non-empty if first, indicates which regions should be preserved from the existing copy.  From this and size, you can work out which sections need to be hole punched).
2) bool clone_omap (only true if first)

You do a clone or clone_range depending on the values of the two arguments 1 || 2 on the first operation as needed and blank out the omap or some/all of the data region as needed.

This section needs to be crystal clear before we can even bother testing it.